sygnal/helper/proxy/proxy_asyncio.py

```diff
@@ -0,0 +1,241 @@
+        self.buffer = b""
+        self.transport: Transport = None  # type: ignore
+        self.proxy_url_parts = proxy_url_parts
+
+        # This future is completed when it is safe to take back control of the
+        # transport (which is also returned to indicate this).
+        self.wait_for_establishment: Future[Transport] = Future()
+
+    def data_received(self, data: bytes) -> None:
+        super().data_received(data)
+        self.buffer += data
+        if b"\r\n\r\n" in self.buffer:
+            # The response headers are terminated by a double CRLF.
+            # NB we want want 'in' instead of 'endswith'
+            #  as no guarantee error page won't come immediately.
+
+            # warning: note this won't work if the remote host talks through
+            # the tunnel first.
+            # (This is OK because:
+            #   - in cleartext HTTP, the client sends the request before the
+            #     server utters a word
+            #   - in TLS, the client talks first by sending a client hello
+            #   - we aren't interested in using anything other than TLS over this
+            #     proxy, anyway
+            # )
+
+            # All HTTP header lines are terminated by CRLF.
+            # the first line of the response headers is the Status Line
+            try:
+                lines = self.buffer.split(b"\r\n")
+                status_line = lines[0]
+                # maxsplit=2 denotes the number of separators, not the № items
+                # StatusLine ← HTTPVersion SP StatusCode SP ReasonPhrase
+                # None of the fields may contain CRLF, and only ReasonPhrase may
+                # contain SP.
+                [http_version, status, reason_phrase] = status_line.split(
+                    b" ", maxsplit=2
+                )
+                logger.debug(
+                    "CONNECT response from proxy: hv=%s, r=%s, rp=%s",
+                    http_version,
+                    status,
+                    reason_phrase,
+                )
+                if status != b"200":
+                    # 200 Successful (aka Connection Established) is what we want
+                    # if it is not what we have, then we don't have a tunnel
+                    logger.error(
+                        "Error from HTTP Proxy"
+                        " whilst attempting CONNECT: %s (%s);"
+                        "aborting connection.",
+                        status,
+                        reason_phrase,
+                    )
+                    self.transport.close()
+                    raise ProxyConnectError(
+                        "Error from HTTP Proxy"
+                        f" whilst attempting CONNECT: {status.decode()}"
+                        f" ({reason_phrase.decode()}); aborting connection."
+                    )
+
+                logger.debug("Ready to switch over protocol")
+
+                self.buffer = None  # type: ignore
```

Comment:
why do we do this, out of interest?

Also: what happens to any bytes in the buffer after the CONNECT response? don't they need to go to the new protocol?

Reply:
> why do we do this, out of interest?

Probably not concerning in the grand scheme of things, but intended to allow the GC to take back the memory spent by the byte buffer.

Would it be less silly looking to use `b""` instead of None?

> Also: what happens to any bytes in the buffer after the CONNECT response? don't they need to go to the new protocol?

In a totally generic world, yes.

But neither HTTP nor TLS allows the server to speak before the client (this is mentioned a little bit up, the context shifted downwards throughout tweaking, I suppose), so the situation is that there are no bytes left over (should I assert this or figure out a way to pass back the theoretical leftovers?).

Reply:
> But neither HTTP nor TLS allows the server to speak before the client

are you sure about this? when I `openssl s_client -connect api.push.apple.com:443`, I get some bytes spat back at me.

Reply:
> are you sure about this? when I `openssl s_client -connect api.push.apple.com:443`, I get some bytes spat back at me.

Me too. I suppose I need to update my assumptions since HTTP/2 … it seems like quite a large rework of HTTP.

I would say I expect TLS will still be fine (and `nc api.push.apple.com 443` doesn't show anything come back before I start typing) but it may just be a matter of time until someone breaks that assumption.

Lucky my new version can handle dangling bytes then … :)